rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,416 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! Client for Key Lease operations.
+
+use std::{sync::Arc, time::Duration};
+
+use tokio::{sync::Mutex, task::JoinHandle};
+
+use crate::leased_lock::{Error, ErrorKind, LeaseObservation, Response, SetCondition, SetOptions};
+use crate::state_store::{self};
+use azure_iot_operations_mqtt::interface::ManagedClient;
+use azure_iot_operations_protocol::common::hybrid_logical_clock::HybridLogicalClock;
+
+/// Key Lease client struct.
+pub struct Client<C>
+where
+    C: ManagedClient + Clone + Send + Sync + 'static,
+    C::PubReceiver: Send + Sync,
+{
+    state_store: Arc<state_store::Client<C>>,
+    key_name: Vec<u8>,
+    lease_holder_name: Vec<u8>,
+    current_fencing_token: Arc<Mutex<Option<HybridLogicalClock>>>,
+    auto_renewal_task: JoinHandle<()>,
```

Comment:
instead of saving a handle to this task, it might be cleaner to have a channel/Notify/cancellation token into the task that can cancel it

Reply:
Yes, that was great advice. Done as we discussed. Thanks.